src/libstd/thread/local.rs

```diff
@@ -364,54 +398,84 @@ pub mod fast {
     impl<T> Key<T> {
         pub const fn new() -> Key<T> {
             Key {
-                inner: UnsafeCell::new(None),
-                dtor_registered: Cell::new(false),
-                dtor_running: Cell::new(false)
+                inner: LazyKeyInner::new(),
+                dtor_state: Cell::new(DtorState::Unregistered),
             }
         }
 
-        pub unsafe fn get(&self) -> Option<&'static UnsafeCell<Option<T>>> {
-            if mem::needs_drop::<T>() && self.dtor_running.get() {
-                return None
+        #[inline]
+        pub unsafe fn get<F: FnOnce() -> T>(&self, init: F) -> Option<&'static T> {
+            match self.inner.get() {
+                Some(val) => Some(val),
+                None => {
+                    if mem::needs_drop::<T>() {
+                        self.try_initialize_drop(init)
+                    } else {
+                        Some(self.try_initialize_nodrop(init))
+                    }
+                }
             }
-            self.register_dtor();
-            Some(&*(&self.inner as *const _))
         }
 
-        unsafe fn register_dtor(&self) {
-            if !mem::needs_drop::<T>() || self.dtor_registered.get() {
-                return
+        // `try_initialize_nodrop` is only called once per fast thread local
+        // variable, except in corner cases where it is being recursively
+        // initialized.
+        //
+        // Macos: Inlining this function causes two `tlv_get_addr` calls to be
+        // performed for every call to `Key::get`.
+        // LLVM issue: https://bugs.llvm.org/show_bug.cgi?id=41722
+        #[inline(never)]
```

Comment:
I think we'll want to remove this `#[inline(never)]` tag if possible from the final codegen. Some initializers are very simple (like the constant zero) and so while `#[cold]` may make sense for branch prediction I think we'll want to leave it up to LLVM to determine whether to inline this

Reply:
Done!

Removing this `#[inline(never)]` puts another `tlv_get_addr` call into the fast path for macos, but that could be fixed in a separate PR.

Reply:
Oh hm I didn't realize it'd add another call on the fast path, but if you think it can be fixed later that's also fine by me!